src/python/pants/base/specs.py

```diff
@@ -60,221 +53,291 @@ def is_directory_shorthand(self) -> bool:
+        file_globs = []
+        dir_literals = []
+        dir_globs = []
+        recursive_globs = []
+        ancestor_globs = []
+        for spec in specs:
+            if isinstance(spec, AddressLiteralSpec):
+                address_literals.append(spec)
+            elif isinstance(spec, FileLiteralSpec):
+                file_literals.append(spec)
+            elif isinstance(spec, FileGlobSpec):
+                file_globs.append(spec)
+            elif isinstance(spec, DirLiteralSpec):
+                if convert_dir_literal_to_address_literal:
+                    address_literals.append(spec.to_address_literal())
+                else:
+                    dir_literals.append(spec)
+            elif isinstance(spec, DirGlobSpec):
+                dir_globs.append(spec)
+            elif isinstance(spec, RecursiveGlobSpec):
+                recursive_globs.append(spec)
+            elif isinstance(spec, AncestorGlobSpec):
+                ancestor_globs.append(spec)
+            else:
+                raise AssertionError(f"Unexpected type of Spec: {repr(spec)}")
+        return Specs(
+            tuple(address_literals),
+            tuple(file_literals),
+            tuple(file_globs),
+            tuple(dir_literals),
+            tuple(dir_globs),
+            tuple(recursive_globs),
+            tuple(ancestor_globs),
+            filter_by_global_options=filter_by_global_options,
+            from_change_detection=from_change_detection,
+        )
 
-    def __str__(self) -> str:
-        return self.glob
+    def __bool__(self) -> bool:
+        return bool(
+            self.address_literals
+            or self.file_literals
+            or self.file_globs
+            or self.dir_literals
+            or self.dir_globs
+            or self.recursive_globs
+            or self.ancestor_globs
+        )
 
-    def to_glob(self) -> str:
-        return self.glob
+    def arguments_provided_description(self) -> str | None:
+        """A description of what the user specified, e.g. 'target arguments'."""
+        specs_descriptions = []
+        if self.address_literals:
+            specs_descriptions.append("target")
+        if self.file_literals or self.file_globs:
+            specs_descriptions.append("file")
+        if self.dir_literals:
+            specs_descriptions.append("directory")
+        if self.dir_globs or self.recursive_globs or self.ancestor_globs:
+            specs_descriptions.append("glob")
+
+        if not specs_descriptions:
+            return None
+        if len(specs_descriptions) == 1:
+            return f"{specs_descriptions[0]} arguments"
+        if len(specs_descriptions) == 2:
+            return " and ".join(specs_descriptions) + " arguments"
+        return ", ".join(specs_descriptions[:-1]) + f", and {specs_descriptions[-1]} arguments"
 
 
 @dataclass(frozen=True)
-class DirLiteralSpec(FilesystemSpec):
-    """A literal dir path, e.g. `some/dir`.
+class SpecsWithoutFileOwners:
```

Comment:
The description of this change is fantastic, and (to me at least) all of the renames and the meanings of the resulting classes are self explanatory.

Except these two: `SpecsWithoutFileOwners` and `SpecsWithOnlyFileOwners`. It's not clear how/why these should be used from their names (in particular: relative to `Specs`), and so docstrings would be good for them I think.

Alternatively: maybe they don't need special types, and instead of `SpecsWithoutFileOwners.from_specs(cls, ..) -> Self`, you could instead have `Specs.without_file_owners(self, ...) -> Self`, which just dropped the irrelevant fields...?

Reply:
We need distinct types because the rule graph - to evaluate `SpecsWithOnlyFileOwners`, we must invoke `Owners` which invokes the rule for `SpecsWithoutFileOwners` -> a graph cycle

Agreed the naming isn't the best. This was the hardest part of the PR for me last night. Any suggestions?

I'll add docstring now

Reply:
@stuhood I will update all relevant callers to directly use `Specs`, rather than `SpecsWithoutFileOwners`. It technically will result in a bigger rule graph, but it's much clearer to readers.